lib/promscrape/discovery/consul/watch.go

```diff
@@ -0,0 +1,237 @@
+	baseQueryArgs  string
+	cancel         context.CancelFunc
+	client         *discoveryutils.Client
+	lastAccessTime uint64
+	// guards services
+	mu                  sync.Mutex
+	nodeMeta            string
+	shouldWatchServices []string
+	shouldWatchTags     []string
+	services            map[string]serviceWatch
+}
+
+// init new watcher and start bachground discovery.
+func newWatchConsul(client *discoveryutils.Client, sdc *SDConfig, dc string) (*watchConsul, error) {
+	// wait time must be less, then fasthttp client deadline - its 1 minute.
+	baseQueryArgs := fmt.Sprintf("?sdc=%s", url.QueryEscape(dc))
+	var nodeMeta string
+	if len(sdc.NodeMeta) > 0 {
+		for k, v := range sdc.NodeMeta {
+			nodeMeta += fmt.Sprintf("&node-meta=%s", url.QueryEscape(k+":"+v))
+		}
+	}
+	if sdc.AllowStale {
+		baseQueryArgs += "&stale"
+	}
+	wc := watchConsul{
+		client:              client,
+		baseQueryArgs:       baseQueryArgs,
+		shouldWatchServices: sdc.Services,
+		shouldWatchTags:     sdc.Tags,
+		services:            make(map[string]serviceWatch),
+	}
+
+	watchServiceNames, _, err := wc.getServiceNames(0)
+	if err != nil {
+		return nil, err
+	}
+	// global context
+	ctx, cancel := context.WithCancel(context.Background())
+	wc.cancel = cancel
+	var syncWait sync.WaitGroup
+	for serviceName := range watchServiceNames {
+		ctx, cancel := context.WithCancel(ctx)
+		syncWait.Add(1)
+		go wc.startWatchService(ctx, serviceName, &syncWait)
+		wc.services[serviceName] = serviceWatch{cancel: cancel}
+	}
+	// wait for first init.
+	syncWait.Wait()
+	go wc.watchForServices(ctx)
+	return &wc, nil
+}
+
+// stops all service watchers.
+func (w *watchConsul) stopAll() {
+	w.mu.Lock()
+	for _, sw := range w.services {
+		sw.cancel()
+	}
+	w.mu.Unlock()
+}
+
+// getServiceNames returns serviceNames and index version.
+func (w *watchConsul) getServiceNames(index uint64) (map[string]struct{}, uint64, error) {
+	sns := make(map[string]struct{})
+	path := fmt.Sprintf("/v1/catalog/services%s", w.baseQueryArgs)
+	if len(w.nodeMeta) > 0 {
+		path += w.nodeMeta
+	}
+	data, newIndex, err := getAPIResponse(w.client, path, index)
+	if err != nil {
+		return nil, index, err
+	}
+	var m map[string][]string
+	if err := json.Unmarshal(data, &m); err != nil {
+		return nil, index, fmt.Errorf("cannot parse services response=%q, err=%w", data, err)
+	}
+	for k, tags := range m {
+		if !shouldCollectServiceByName(w.shouldWatchServices, k) {
+			continue
+		}
+		if !shouldCollectServiceByTags(w.shouldWatchTags, tags) {
+			continue
+		}
+		sns[k] = struct{}{}
+	}
+	return sns, newIndex, nil
+}
+
+// listen for new services and update it.
+func (w *watchConsul) watchForServices(ctx context.Context) {
+	ticker := time.NewTicker(*SDCheckInterval)
+	defer ticker.Stop()
+	var index uint64
+	for {
+		select {
+		case <-ctx.Done():
+			w.stopAll()
+			return
+		case <-ticker.C:
+			if fasttime.UnixTimestamp()-atomic.LoadUint64(&w.lastAccessTime) > uint64(SDCheckInterval.Seconds())*2 {
```

Comment:
There is no need in using `fasttime` here, since this loop is executed rarely. It looks like it would be better from readability point of view to use functions from standard `time` package here.

Reply:
changed